File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, ArrayList<String>> anagrams = new HashMap<>();
        for (String s : strs) {
            char[] c = s.toCharArray();
            Arrays.sort(c);
            String sNew = new String(c);
            if (!anagrams.containsKey(sNew)) anagrams.put(sNew, new ArrayList<>());
            anagrams.get(sNew).add(s);
        }
        return new ArrayList(anagrams.values());
    }
}

File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        for (string& s : strs) {
            string temp = s;
            sort(temp.begin(), temp.end());
            groups[temp].emplace_back(s);
        }
        
        vector<vector<string>> output;
        output.reserve(groups.size());
        for (auto& item : groups) output.emplace_back(item.second);
        return output;
    }
};
